### PR TITLE
Null-safety and DIO 4.0.0 support

### LINKED ISSUE #67: Migrate to null safety
## Description

Dart has recently obtained the ability to enable sound null safety in Dart code. [Here](https://dart.dev/null-safety) is the overview of the feature. It is currently in beta.

From the [migration guide](https://dart.dev/null-safety/migration-guide) we see the following notice:

> **Migrating an app is technically the same as migrating a package.** Before migrating an app, consider waiting until null safety is in a stable release and all your dependencies are ready.

This means that it's generally recommended to migrate if developing a Dart package. Thus, it would seem beneficial to implement null safety features in `http_mock_adapter`.

### PR DESCRIPTION
**Pull request by kuhnroyal**
## Description

This PR adds null-safety and DIO 4.0.0 support. Fixes #67

This PR is based on #82 so that one needs to be merge first, afterwards I can rebase. That will make the diff clearer. 

This should be published as `1.0.0` or `1.0.0-nullsafety.1` preview.

 CC @LukaGiorgadze @GiorgiBeriashvili @campfire5

### CODE REVIEW

**Comment by GiorgiBeriashvili** [comment]
Sounds exciting! #82 is even more important now. I will focus on it immediately and will provide feedback soon.

**Comment by magicleon94** [comment]
Any news on this? I have a migration blocked by this :(

**Comment by kuhnroyal** [comment]
You can use an override pointing to this branch for now.

**Comment by GiorgiBeriashvili** [comment]
I merged the PR (#82) the changes were based on.

> This should be published as 1.0.0 or 1.0.0-nullsafety.1 preview.

Could you please explain why is that the case? I was thinking on either `0.2.0` or `0.3.0`.

**Comment by kuhnroyal** [comment]
> I merged the PR (#82) the changes were based on.
> 
> > This should be published as 1.0.0 or 1.0.0-nullsafety.1 preview.
> 
> Could you please explain why is that the case? I was thinking on either `0.2.0` or `0.3.0`.

Well this a breaking change and according to semantic versioning this should be a major version. And it would automatically be pulled in for anyone with `^0.1.0` in the pubspec where as `1.0.0` would not get automatically updated.

I rebased it on develop.

**Comment by GiorgiBeriashvili** [comment]
> > I merged the PR (#82) the changes were based on.
> > > This should be published as 1.0.0 or 1.0.0-nullsafety.1 preview.
> > 
> > 
> > Could you please explain why is that the case? I was thinking on either `0.2.0` or `0.3.0`.
> 
> Well this a breaking change and according to semantic versioning this should be a major version. And it would automatically be pulled in for anyone with `^0.1.0` in the pubspec where as `1.0.0` would not get automatically updated.
> 
> I rebased it on develop.

Understandable, though the package is still under `1.0.0`, in which case it could be decided how the version will be raised: https://dart.dev/null-safety/migration-guide#package-version

**Comment by kuhnroyal** [comment]
@GiorgiBeriashvili That is up to you, since it is "only" a test dependency either way is probably fine.

File: .github/workflows/dart-ci.yml
Comment (GiorgiBeriashvili) [inline .github/workflows/dart-ci.yml:22]: Is this still necessary?

Dart SDK stable has reached `2.12.0` version with the release of Flutter `2.0` I think, which implies that `beta` might not be necessary anymore.

File: example/main.dart
Comment (GiorgiBeriashvili) [inline example/main.dart:105]: A bit of nitpicking but maybe passing constant declared at the beginning of the file - `path` will make more sense in this case just for the sake of consistency, though it essentially changes nothing.

File: pubspec.yaml
Comment (GiorgiBeriashvili) [inline pubspec.yaml:5]: I think this should change, but not sure to which version as of yet. Though, I personally am thinking on `0.2.0`. I will ask my colleagues for confirmation regarding the version.

File: test/interceptors/dio_interceptor_test.dart
Comment (GiorgiBeriashvili) [inline test/interceptors/dio_interceptor_test.dart:49]: A bit of nitpicking but maybe passing constant declared at the beginning of the file - `path` will make more sense in this case just for the sake of consistency, though it essentially changes nothing.

**Review by GiorgiBeriashvili — CHANGES_REQUESTED** [review]
I tested the changes and everything seems to be working as expected. I simply left small comments, let me know what you think. I will try to advocate to merge the changes quickly.

Thank you, @kuhnroyal

File: test/adapters/dio_adapter_test.dart
Comment (GiorgiBeriashvili) [inline test/adapters/dio_adapter_test.dart:40]: A bit of nitpicking but maybe passing constant declared at the beginning of the file - `path` will make more sense in this case just for the sake of consistency, though it essentially changes nothing.

**Comment by kuhnroyal** [comment]
I addressed the paths and CI change. The version is up to you.
Not exactly sure why the build fails now, need to look into this.

**Comment by kuhnroyal** [comment]
Ok dio was updated again. This might happen again until dio 4.0.0 is released.

Also I noticed that the format test in CI doesn't fail - `set-exit-if-changed` not working . See https://github.com/dart-lang/sdk/issues/44582

**Comment by kuhnroyal** [comment]
Update to latest Dio 4.0.0-prev1 and implemented the required interceptor API changes.

**Comment by LukaGiorgadze** [comment]
> Update to latest Dio 4.0.0-prev1 and implemented the required interceptor API changes.

Thanks @kuhnroyal , i added @GiorgiBeriashvili and @campfire5 as a (re-)reviewers.

**Review by GiorgiBeriashvili — APPROVED** [review]

**Review by GiorgiBeriashvili — APPROVED** [review]

**Review by campfire5 — APPROVED** [review]

**Comment by GiorgiBeriashvili** [comment]
@kuhnroyal Could you please resolve the conflicts (two PRs have been merged in the meantime) and change the package's version to `0.2.0` in **pubspec.yaml**?

Thereafter, after re-reviewing once again, merging will most probably be in order.

**Comment by kuhnroyal** [comment]
On it.

**Review by GiorgiBeriashvili — APPROVED** [review]

File: lib/src/request.dart
Comment (campfire5) [inline lib/src/request.dart:42]: why make it on one line ?

Comment (kuhnroyal) [reply lib/src/request.dart]: Yea this got formatted to 150 line length in the merge, my bad.

**Review by GiorgiBeriashvili — APPROVED** [review]

File: lib/src/adapters/dio_adapter_mockito.dart
Comment (campfire5) [inline lib/src/adapters/dio_adapter_mockito.dart:47]: I need to look at it better, I will do it today

**Review by campfire5 — APPROVED** [review]

**Comment by LukaGiorgadze** [comment]
@kuhnroyal First of all, thank you for your time and effort! Thank you for being actively contributing in open source world!

We checked this PR and approved, but one small thing is that some of PRs merged before and we have conflicts. I would like to resolve it for you, but i'm worrying to change something else.

Please, help us resolving this conflicts and we will merge it ASAP.

Thank you again!

**Comment by kuhnroyal** [comment]
You seemed to have force pushed over the `develop` and `main` when trying to release 0.1.6. This is really bad practice, now all forks are out of sync. The last shared ancestor commit is on the the `0.1.4` tag (19310519550fc6402eb760ee5f3ef0757d187b89).

The commit of the `0.1.5` tag (https://github.com/lomsa-dev/http-mock-adapter/commit/ff0b5b1c9d976e774002f3176fa0b6acd193c715) is not even part of the repository anymore.

I strongly recommend reverting all the changes and resetting `main` and `develop` to before the force-push.
Additionally something got messed up during the rebase, the conflicts make no sense.

**Comment by LukaGiorgadze** [comment]
> You seemed to have force pushed over the `develop` and `main` when trying to release 0.1.6. This is really bad practice, now all forks are out of sync. The last shared ancestor commit is on the the `0.1.4` tag ([1931051](https://github.com/lomsa-dev/http-mock-adapter/commit/19310519550fc6402eb760ee5f3ef0757d187b89)).
> 
> The commit of the `0.1.5` tag ([ff0b5b1](https://github.com/lomsa-dev/http-mock-adapter/commit/ff0b5b1c9d976e774002f3176fa0b6acd193c715)) is not even part of the repository anymore.
> 
> I strongly recommend reverting all the changes and resetting `main` and `develop` to before the force-push.
> Additionally something got messed up during the rebase, the conflicts make no sense.

@kuhnroyal `develop` is reverted and ready to merge. After this merge, we need to resolve conflicts between `develop` and `main`.

**Pull request merged**
